test/correctness/atomics.cpp

```diff
@@ -0,0 +1,1223 @@
+        } break;
+    }
+
+    Buffer<int> correct(hist_size);
+    correct.fill(0);
+    for (int i = 0; i < img_size; i++) {
+        int idx = (i*i) % hist_size;
+        correct(idx) = correct(idx) + 1;
+    }
+
+    // Run 10 times to make sure race condition do happen
+    for (int iter = 0; iter < 10; iter++) {
+        Buffer<int> out = consumer.realize(hist_size);
+        for (int i = 0; i < hist_size; i++) {
+            check(__LINE__, out(i), correct(i));
+        }
+    }
+}
+
+void test_async_tuple(const Backend &backend) {
+    int img_size = 10000;
+    int hist_size = 7;
+
+    Func producer0, producer1, consumer0, consumer1;
+    Var x;
+    RDom r(0, img_size);
+    RDom rh(0, hist_size);
+
+    producer0(x) = (x*x) % hist_size;
+    producer1(x) = ((x+1)*(x-1)) % hist_size;
+
+    consumer0(x) = Tuple(0, 0);
+    consumer0(producer0(r)) += Tuple(1, 1);
+    consumer0(producer1(r)) += Tuple(1, 1);
+
+    consumer1(x) = Tuple(0, 0);
+    consumer1(clamp(consumer0(rh)[0], 0, 2*img_size)) += Tuple(1, 1);
+
+    consumer0.compute_root().async();
+    producer0.compute_root().async().parallel(x);
+    producer1.compute_root().async().parallel(x);
+    consumer1.compute_root();
+    switch(backend) {
+        case Backend::CPU: {
+            consumer0.update(0)
+                     .atomic(true /*override_associativity_test*/)
+                     .parallel(r);
+            consumer0.update(1)
+                     .atomic(true /*override_associativity_test*/)
+                     .parallel(r);
+            consumer1.update()
+                     .atomic()
+                     .parallel(rh);
+        } break;
+        default: {
+            _halide_user_assert(false) << "Unsupported backend.\n";
+        } break;
+    }
+
+    Buffer<int> correct_consumer0(hist_size);
+    Buffer<int> correct_consumer1(2 * img_size);
+    correct_consumer0.fill(0);
+    correct_consumer1.fill(0);
+    for (int i = 0; i < img_size; i++) {
+        int idx = (i*i) % hist_size;
+        correct_consumer0(idx) = correct_consumer0(idx) + 1;
+    }
+    for (int i = 0; i < img_size; i++) {
+        // Halide's modulo behaves differently compared to C's modulo.
+        int idx = Halide::Internal::mod_imp(((i+1)*(i-1)), hist_size);
+        correct_consumer0(idx) = correct_consumer0(idx) + 1;
+    }
+    for (int i = 0; i < hist_size; i++) {
+        int idx = correct_consumer0(i);
+        correct_consumer1(idx) = correct_consumer1(idx) + 1;
+    }
+
+    // Run 10 times to make sure race condition do not happen
+    for (int iter = 0; iter < 10; iter++) {
+        Realization out = consumer1.realize(2 * img_size);
+        Buffer<int> out0 = out[0];
+        Buffer<int> out1 = out[1];
+        for (int i = 0; i < 2 * img_size; i++) {
+            check(__LINE__, out0(i), correct_consumer1(i));
+            check(__LINE__, out1(i), correct_consumer1(i));
+        }
+    }
+}
+
+int main(int argc, char **argv) {
```

Comment:
This test seems large. How long is the runtime? (Wondering if it would be sensible to split it into several tests)

Reply:
See Andrew's comment: around a minute.

Reply:
I got it to go substantially faster than that by trimming down the number of iterations for the code that tries to make it more likely to detect race conditions.